lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,375 @@
+
+        # Something invalid in the card, e.g. verify declined when linking card to customer.
+        'INVALID_CARD_DATA' => STANDARD_ERROR_CODE[:processing_error],
+        'CARD_EXPIRED' => STANDARD_ERROR_CODE[:expired_card],
+        'VERIFY_CVV_FAILURE' => STANDARD_ERROR_CODE[:incorrect_cvc],
+        'VERIFY_AVS_FAILURE' => STANDARD_ERROR_CODE[:incorrect_zip],
+        'CARD_DECLINED' => STANDARD_ERROR_CODE[:card_declined],
+        'UNAUTHORIZED' => STANDARD_ERROR_CODE[:config_error]
+      }
+
+      # The `login` key is the client_id (also known as application id) in the dev portal
+      # Get it after you create a new app: https://connect.squareup.com/apps/
+      # The `password` is the access token (personal or OAuth)
+      # The `location_id` must be fetched initially https://docs.connect.squareup.com/articles/processing-payment-rest/
+      # The `test` indicates if these credentials are for sandbox or production (money moving) access
+      def initialize(options={})
+        requires!(options, :login, :password, :location_id, :test)
+        @client_id = options[:login].strip
+        @bearer_token = options[:password].strip
+        @location_id = options[:location_id].strip
+
+        super
+      end
+
+      # To create a charge on a card using a card nonce:
+      #     purchase(money, card_nonce, { ...create transaction options... })
+      #
+      # To create a customer and save a card (via card_nonce) to the customer:
+      #     purchase(money, card_nonce, {customer: {...params hash same as in store() method...}, ...})
+      #   Note for US and CA, you must have {customer: {billing_address: {zip: xxxxx}}} which passes
+      #   AVS to store a card.
+      #   Note this always creates a new customer, so it may make a duplicate customer if this card was
+      #   associated to another customer previously.
+      #
+      # To use a customer's card on file:
+      #     purchase(money, nil, {customer: {id: 'customer-id', card_id: 'card-id'}})
+      #   Note this does not update any fields on the customer.
+      #
+      # To use a customer, but link a new card to the customer:
+      #     purchase(money, card_nonce, {customer: {id: 'customer-id'}})
+      #   Note this does not update any other fields on the customer.
+      #
+      # As this may take multiple requests, it returns a MultiResponse.
+      def purchase(money, card_nonce, options={})
+        raise ArgumentError('money required') if money.nil?
+        if card_nonce.nil?
+          requires!(options, :customer)
+          requires!(options[:customer], :card_id, :id)
+        end
+        if card_nonce && options[:customer] && options[:customer][:card_id]
+          raise ArgumentError('Cannot call with both card_nonce and options[:customer][:card_id], choose one.')
+        end
+
+        post = {}.merge(options.slice(:buyer_email_address, :delay_capture, :note, :reference_id))
```

Comment:
I think `post = options.slice(:buyer_email_address, :delay_capture, :note, :reference_id)` should be sufficient